### ISSUE TITLE
How to add all untracked files with git.add?

### ISSUE THREAD

**Issue by keshin** [isomorphic-git/isomorphic-git#187]
Hi,
I'm looking for a way init a git repo, add all existing files, and commit them, while if I understand correctly, the git.add function needs me to give all of the files explicitly.

Is there a way that I can add all of the files in one or two command?

**Comment by wmhilton** [comment]
**TLDR**
```js
const globby = require('globby');
const paths = await globby(['./**', './**/.*'], { gitignore: true });
for (const filepath of paths) {
    await git.add({ fs, dir, filepath });
}
```
**Long answer**
Good question! I didn't add an "add all" function because you almost always want to make an exception or two, and the general category of "get me a list of all the files matching some pattern" is pretty well covered by existing modules on npm. But a good example to demonstrate it would be something I should add that to the FAQ. (\<looks at website\> Wait, I don't have an FAQ yet? I should make one. https://github.com/isomorphic-git/isomorphic-git/issues/191)

How's about this recipe? Here I use [globby](https://npmjs.org/package/globby) because it has a super nifty feature: it can understand `.gitignore` files! (It's written for node... I haven't tested whether globby is compatible with BrowserFS.)

```js
#!/usr/bin/env node
const globby = require('globby');
const git = require('isomorphic-git');
const fs = require('fs');
const dir = process.cwd();

(async () => {
    await git.init({ fs, dir });
    const paths = await globby(['./**', './**/.*', '!node_modules'], { gitignore: true });
    for (const filepath of paths) {
        await git.add({ fs, dir, filepath });
    }
    await git.config({ fs, dir, path: 'user.name', value: 'Some User' });
    await git.config({ fs, dir, path: 'user.email', value: 'someuser@example.com' });
    await git.commit({ fs, dir, message: 'Initial commit' })
})();
```

I'll break this example down to make things clearer:

```js
(async () => {
```

Until await is allowed at the module level, we have to wrap await statements in an immediately invoked anonymous async function, or else use the `.then()` syntax for dealing with promises.

```js
    const paths = await globby(['./**', './**/.*', '!node_modules'], { gitignore: true });
```

This tells globby to:
- `./**` = match all regular files
- .`/**/.*` = match all "hidden" dot files, like .babelrc or .travis.yml
- `!node_modules` = exclude the node_modules folder
- `{ gitignore: true} ` = also respect the patterns in .gitignore files

```js
    for (const filepath of paths) {
        await git.add({ fs, dir, filepath });
    }
```
This is a loop that calls `git.add` for each file.

```js
    await git.config({ fs, dir, path: 'user.name', value: 'Some User' });
    await git.config({ fs, dir, path: 'user.email', value: 'someuser@example.com' });
    await git.commit({ fs, dir, message: 'Initial commit' })
```

`git.commit` will throw an error if no author information is available. Here I'm saving it to the repo's local config file so that you won't have to enter that information each time you commit. But you can alternatively pass that information as an argument, e.g.

```js
        await git.add({ fs, dir, filepath, author: {name: 'Some User', email: 'someuser@example.com' }})
```

Whew! This turned into a longer answer than I expected. Hope it helps!!

**Comment by jcubic** [comment]
I've look at globby index.js in jsdeliver cdn and it's node module (with node dependencies with require) it will not work in browser.

But you can use this function based on this https://gist.github.com/slav123/8118962 it only use fs.

```javascript
function listFiles(startDir, usePath, callback) {
    if (arguments.length === 2 && typeof arguments[1] === 'function') {
        callback = usePath;
        usePath = false;
    }
    //Hold onto the array of items
    var parsedDirectory = [];
    //start reading a list of whats contained
    fs.readdir(startDir, function(err, dirList) {
        if (usePath) {
            startDir = fs.realpathSync(startDir);
        }
        if (err) {
            return callback(err);
        }
        //keep track of how deep we need to go before callback
        var listlength = dirList.length;
        if (!listlength) {
            return callback(null, parsedDirectory);
        }
        //loop through the directory list
        dirList.forEach(function(file) {
            file = (startDir === '/' ? '' : startDir) + '/' + file;
            fs.stat(file, function(err, stat) {
                //note the directory or file
                if (stat && stat.isFile()) parsedDirectory.push(file);
                //recursive if this is a directory
                if (stat && stat.isDirectory()) {
                    //recurse
                    traverseDirectory(file, function(err, parsed) {
                        // read this directory into our output
                        parsedDirectory = parsedDirectory.concat(parsed);
                        //check to see if we've exhausted our search
                        if (!--listlength) {
                            callback(null, parsedDirectory);
                        }
                    });
                } else {
                    //check to see if we've exhausted the search
                    if (!--listlength) {
                        callback(null, parsedDirectory);
                    }
                }
            });
        });
    });
}
```

**Comment by keshin** [comment]
It helped definitely! Thanks a lot!

**Comment by wmhilton** [comment]
Finally added this to the FAQ (which finally exists): https://isomorphic-git.org/docs/en/faq#how-to-add-all-untracked-files-with-gitadd

**Comment by Ehesp** [comment]
I just stumbled across this, but I've been doing `git.add('.')` and seems to be working ok. Do we need to iterate over each file?

**Comment by jcubic** [comment]
@Ehesp this discussion is a few years old, if it works it means that it works.

**Comment by supnate** [comment]
It works for me too: `await git.add({ fs, dir: prjRoot, filepath: '.' });`
